core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -1082,4 +1085,33 @@ public void close() {
 
         return new WriteResult(new IndexResponse(shardId.getIndex(), request.type(), request.id(), request.version(), created), operation.getTranslogLocation());
     }
+
+    protected final void processAfter(boolean refresh, IndexShard indexShard, Translog.Location location) {
+        if (refresh) {
+            try {
+                indexShard.refresh("refresh_flag_index");
+            } catch (Throwable e) {
+                // ignore
+            }
+        }
+
+        if (indexShard.getTranslogDurability() == Translog.Durabilty.REQUEST && location != null) {
+            indexShard.sync(location);
+        }
+        if (indexShard.flushPending()) {
+            threadPool.executor(ThreadPool.Names.FLUSH).execute(() -> {
```

Comment:
this can create a storm of tasks pending to be executed on the thread pool. For example, if a flush is needed, and we are doing 5k indexing requests per second, until the thread get scheduled, and the flush actually executes to "cleanup" the flushPending (can take time), there will be thousands of tasks pending on the thread pool, eventually executing on a thread and potentially not doing anything. If active indexing is happening all the time and flushing is continuously needed, we will have an exponential number of thread switches to execute a flush.

Might it make sense to have an AtomicBoolean like behavior encapsulated in IndexShard? an "executing flush" like semantics?